### DISCUSSION TITLE
Fatal exceptions using RetryTopicConfiguration

### DISCUSSION THREAD

**Discussion by einarjohnson** [spring-projects/spring-kafka#1820] [Q&A] [answered]
Hi,
Is it possible to set a retriable topic configuration up in such a way that certain fatal exceptions go straight to the DLQ topic/handler?

**Comment by garyrussell** [comment, marked as answer]
For normal retry processing, see https://docs.spring.io/spring-kafka/docs/current/reference/html/#seek-to-current

>Starting with version 2.3, the SeekToCurrentErrorHandler considers certain exceptions to be fatal, and retries are skipped for such exceptions; the recoverer is invoked on the first failure. The exceptions that are considered fatal, by default, are:

>DeserializationException

>MessageConversionException

>ConversionException

>MethodArgumentResolutionException

>NoSuchMethodException

>ClassCastException

>since these exceptions are unlikely to be resolved on a retried delivery.

>You can add more exception types to the not-retryable category, or completely replace the map of classified exceptions. See the Javadocs for SeekToCurrentErrorHandler.setClassifications() for more information, as well as those for the spring-retry BinaryExceptionClassifier.

See

https://github.com/spring-projects/spring-kafka/blob/ad1ae60cd8fe98333b16696ba4c37aa5ad07a43a/spring-kafka/src/main/java/org/springframework/kafka/listener/FailedRecordProcessor.java#L74-L92

to replace the defaults, and

https://github.com/spring-projects/spring-kafka/blob/ad1ae60cd8fe98333b16696ba4c37aa5ad07a43a/spring-kafka/src/main/java/org/springframework/kafka/listener/FailedRecordProcessor.java#L162-L181

**Reply by garyrussell** [reply]
@tomazfernandes this is for normal error handling; is there a way to skip the intermediate retry topics with the non-blocking retry?

**Reply by tomazfernandes** [reply]
Hi @garyrussell, sure there is, it's just a matter of specifying the exceptions to exclude. The excluded exceptions will go straight to the DLT.

@einarjohnson, please take a look at https://docs.spring.io/spring-kafka/reference/html/#exception-classifier and let us know if this is what you're looking for.

Thanks!

**Reply by garyrussell** [reply]
I just looked at the code and it is supported via the `DestinationTopicPropertiesFactory` - you pass in a classifier there and if an exception is not retryable, it goes straight to the DLT topic.

Test case here.

https://github.com/spring-projects/spring-kafka/blob/ad1ae60cd8fe98333b16696ba4c37aa5ad07a43a/spring-kafka/src/test/java/org/springframework/kafka/retrytopic/DestinationTopicPropertiesFactoryTests.java#L71-L92

However, I don't see a way to set the `classifierBuilder` in `RetryTopicConfigurationBuilder` used here

https://github.com/spring-projects/spring-kafka/blob/ad1ae60cd8fe98333b16696ba4c37aa5ad07a43a/spring-kafka/src/main/java/org/springframework/kafka/retrytopic/RetryTopicConfigurationBuilder.java#L331-L336

But I might be missing something. @tomazfernandes ?

**Reply by garyrussell** [reply]
Ah - I see; set on the annotation itself.

But, shouldn't we provide a way to set the `classifierBuilder` on the `RetryTopicConfigurationBuilder` for when programmatic configuration is used?

**Reply by tomazfernandes** [reply]
@garyrussell, for programmatic configuration we can use something like this:

```
@Bean
public RetryTopicConfiguration myRetryTopic(KafkaTemplate<String, MyOtherPojo> template) {
    return RetryTopicConfigurationBuilder
            .newInstance()
            .notRetryOn(MyDontRetryException.class)
            .traverseCauses()
            .create(template);
}
```

**Reply by garyrussell** [reply]
Doh - sorry; got it now.

Great work, once again!

**Reply by tomazfernandes** [reply]
Thanks a lot @garyrussell! Glad to see users bringing questions :)

**Reply by einarjohnson** [reply]
@garyrussell and @tomazfernandes , thanks for your help here. I will try the `notRetryOn` on the builder.
